Benchmarking dependency adoption (#36)
epiverse-trace/epiverse-trace.github.io: posts/benchmarking_design_decisions/index.qmd

author:
  - name: "James Mba Azam"
    orcid: "https://orcid.org/0000-0001-5782-7330"
date: last-modified
categories: [benchmarking, R, tidyverse, cli, R packages, software architecture]
format:
  html:
    toc: true
execute: 
  cache: true
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  cache = TRUE,
  message = FALSE
)
```

R package developers are constantly faced with design decisions about user
friendliness. Such design decisions often include checking that inputs are
rightly specified and how to fix or signal issues arising from intermediate
calculations or processes within a function (condition handling).

These design decisions often result in taking on dependencies. Dependencies are
external packages that enable a function to work properly. For example, the
`tidyverse` is a collection of packages that are often used together to
make data analysis easier. The `tidyverse` is a dependency for many
packages that are built on top of it.

Dependencies should not be adopted blindly without carefully considering the
overhead they introduce. Overhead here refers to the additional work that a
package must do to make the dependency work. In this article, we will

> demonstrate how to use (micro)benchmarking[^1] to guide the decision to take on

Reviewer: Looks like you opted to use the bench package in the end so this note should probably be removed.

Author: Not sure I understand. I'm not yet referring to any package in this paragraph.

Reviewer: Oh, I misunderstood you intent then. I took this was a foreshadowing of the fact you were going to use the microbenchmark package.
I now understand was to refer to the concept. Feel free to disregard this comment then.

dependencies.

For this demonstration, we will look at condition handling and signalling,
that is, what to do when problems arise during code execution [^2].

> Condition signalling in R is classified into errors and warnings. Errors refer

Reviewer: We maybe don't need to go into a technical discussion here but just flagging that this is not entirely accurate. Messages are conditions as well and R allows you to create conditions of any class.

Author: Thanks. I've added that.

to cases where something fatal happens during code execution so that it halts,
for example, when a function is run without specifying the required arguments.

```{r}
try(rnorm())
```

Warnings are used to handle less fatal issues with code execution that do not
need to halt the process.

```{r}
c(1:2) > c(1:3)
```

Base R ships with functionalities for condition handling and signalling. There is

> `stop()` for signaling errors and `warning()` for throwing warnings. These base R
> functions are, however, are not flexible enough for advanced formatting: colour,
> progress bar, contextual pluralisation, and so forth.

Reviewer: Nitpicking here but: base R is flexible enough for this and this is precisely how cli manages to build on top of it to add extra features.

What about something like:

```suggestion
`stop()` for signaling errors and `warning()` for throwing warnings. These base R
functions do not include desirable advanced formatting features: colour,
progress bar, contextual pluralisation, and so forth.
```

Author: Awesome. Thanks.

> Package developers often have to adopt external dependencies for flexibility
> in condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features

Reviewer: ```suggestion
Package developers often have to adopt external dependencies for feature-rich condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
```

that will not be discussed here, but of note is its `cli_abort()` and
`cli_warn()` functions for signalling errors and warnings respectfully.

`cli` is a great candidate for flexibility in condition signalling but one
might naturally wonder if this comes at a cost. Here,
we will investigate this through a benchmarking exercise against base R's
`warning()` function. The results of this exercise will provide us with insights

> on whether its worth taking on `cli` for its flexibility as a trade-off for its

Reviewer: ```suggestion
on whether its worth taking on `cli` for its features as a trade-off for its
```

the run times, `purrr`, `tidyr`, and `dplyr` for running the functions and
wrangling the results, and and `ggplot2` for visualizing the results.

```{r}
library(bench)
library(cli)
library(ggplot2)
library(purrr)
library(dplyr)
library(tidyr)
```

Let's define a function that takes an argument `n` for the number of warnings
to throw and `pkg` for the type of package to use.

```{r}
warn_Ntimes <- function(n, pkg) {
  warning_msg <- "x must be an integer"
  switch(pkg,
    base = for (i in 1:n) {
      warning(warning_msg)
    },
    cli = for (i in 1:n) {
      cli_warn(warning_msg)
    }
  )
}
```

Let's test our function to see if it works as expected.

```{r}
warn_Ntimes(3, "base")
warn_Ntimes(3, "cli")
```

Now, we'll consider scenarios where a function throws 1, 5, 10, 15, 20, and
100 warnings using base R and `cli`.

```{r warning=FALSE}
# Number of warnings to throw
warnings <- c(1, 5, 10, 15, 20, 100)
```

Let's run benchmarks over the different scenarios and store results
in a data.frame.

```{r warning=FALSE}
bnmark_list <- warnings |>
  map(
    \(x) mark(
      "cli::cli_warn()" = warn_Ntimes(x, pkg = "cli"),
      "base::warning()" = warn_Ntimes(x, pkg = "base")
    )

>   )

Reviewer: Since you're using bench, it's probably simpler to use `press()` directly. It should be something like:

```r
bench::press(
  fct = c("cli", "base"),
  times = c(1, 5, 10, 15, 20, 100),
  warn_Ntimes(times, fct)
)
```

https://bench.r-lib.org/reference/press.html

Author: Wow, that's very convenient. Thanks.

Their benchmarks are available in the cli documentation [^4]. So, should we
be worried about the speed of `cli`? Well, it depends on the context.

Most developers might argue that this is an optimisation overkill[^5]. However,
we argue that it is important to consider speed differences in context.
In the case of simple printing, the speed difference is
negligible yet disruptive and somewhat painful. However, in the grand scheme
of things, this might be nothing compared to much slower processes that
need more attention. In those cases, the developer might want to consider other
optimisation strategies such as profiling [^6]. The essence of this experiment
is to demonstrate that benchmarking is one way to help developers make design
decisions.

## Conclusion

In designing R package infrastructure with "heavy" dependencies, it might
sometimes be necessary to test your intuition about their overhead with
experiments. Here, we have demonstrated how benchmarking is one way to achieve
this. We show how a simple decision to use `cli::cli_warn()` to handle
warnings could come at the cost of a tiny loss in speed, which is worth
considering in its context.

> 

Reviewer: Would it make sense to link to https://r-pkgs.org/dependencies-mindset-background.html#sec-dependencies-pros-cons for a longer discussion on why every dependency taking decision is unique and how to take context into account?

Author: I've linked it now.

---
title: "Using benchmarking to guide the adoption of dependencies in R packages"
author:
  - name: "James Mba Azam"
    orcid: "https://orcid.org/0000-0001-5782-7330"
date: last-modified
categories: [benchmarking, R, tidyverse, cli, R packages, software architecture]
format:
  html:
    toc: true
execute: 
  cache: true
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  cache = TRUE,
  message = FALSE
)
```

R package developers are constantly faced with design decisions about user
friendliness. Such design decisions often include checking that inputs are
rightly specified and returning feedback early, deciding on how to deal with
issues arising during code execution within a function, and how to
structure outputs.

> These design decisions often result in taking on dependencies. Dependencies are

Reviewer 2: Not sure I agree with this definition of what a dependency is. A function could work properly without calling code from an external package, but it might be better to not rewrite that code, or the dependency could provide a more feature-rich function than base R. I would go with the more general definition used here: https://r-pkgs.org/dependencies-mindset-background.html

Author: Thanks. I've revised it here d08b3d6254cbb31a44d4e38ad2fd42c617d0bd80.

external packages that enable a function to work properly. For example, the

> `tidyverse` is a collection of packages that are often used together to

Reviewer 2: I like the use of the tidyverse as an example as many readers will be familiar with it. However, it might not be the best example for this as it is not a dependency that in itself provides any functionality, but rather bundles functionality from other packages. I think either way the reader will understand what is being explained. 

Might also be good to hyperlink one of the "`tidyverse`" to their website.

Author: Yeah, I initially wanted to leave it as the generic bundle but I've narrowed it down to one example using ggplot2 here d08b3d6254cbb31a44d4e38ad2fd42c617d0bd80.

make data analysis easier. The `tidyverse` is a dependency for many
packages that are built on top of it.

Dependencies should only be adopted after carefully considering their pros and

> cons. In this article, we will demonstrate how to use (micro)benchmarking[^1]
> to guide the decision to take on dependencies that might lead to slower
> processes. In data analytics pipelines, speed is often a key consideration,

Reviewer 2: ```suggestion
cons. In this article, we will demonstrate how to use (micro)benchmarking[^1]
to guide decision-making on dependencies that might lead to slower
processes. In data analytics pipelines, speed is often a key consideration,
```

especially when working with large datasets. Therefore, it is important to
consider how the new dependency might impact the speed of an existing
process.

For this demonstration, we will look at condition signalling,
that is, how to signal to a user that an issue has arisen during code
execution [^2]. Condition signalling in R is classified into errors,
warnings, and messages. Errors refer
to cases where something fatal happens during code execution so that it halts,
for example, when a function is run without specifying the required arguments.

```{r}
try(rnorm())
```

Warnings are used to handle less fatal issues with code execution that do not
need to halt the process. For example, when you compare two vectors of
different lengths, R will throw a warning but will still return a result.

```{r}
c(1:2) > c(1:3)
```

Messages are used to provide useful information about processes. For
example, when you load a package, R will throw a message to indicate that the

> package has been loaded.

Reviewer 2: I think this only happens when there are namespace conflicts, and otherwise the package loads silently (assuming the package author didn't write a message to be printed in `.onLoad()`).

Author: True. I'll make the example more specific to this package.

Author: Thanks @pratikunterwegs. I'll add the qmd style option in the chunk.

```{r}
library(dplyr)
```

Base R ships with functionalities for condition signalling. There is
`stop()` for signaling errors, `warning()` for throwing warnings, and
`message()` for throwing messages. These base R
functions are, however, not flexible enough for advanced
formatting: colour, progress bar, contextual pluralisation, and so forth.

Package developers often have to adopt external dependencies for flexibility
in condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
that will not be discussed here, but of note is its `cli_abort()` and
`cli_warn()` functions for signalling errors and warnings respectfully.

`cli` is a great candidate for flexibility in condition signalling but one

> might naturally wonder if this comes at a cost. Here,

Reviewer 2: Before reading this post, I would have naively assumed that {cli} was performant enough not to care about understanding/quantifying its speed. Therefore, it might be worth mentioning that benchmarking can be used to check whether a introduced function from an external package slows down the code even when it is not expected. This point about unexpected slowdowns/bottlenecks might be better placed in the conclusion.

Author: Yes, that makes sense. I've made the change here a2a5ddf1625a16ddbb03aacdf1ffae0dc55635a6.

on whether its worth taking on `cli` for its features as a trade-off for its
slower speed.

## Experiment

We will consider different scenarios of number of warnings thrown to
tease out the relationship between speed of the function and number of warnings
thrown in a function with and without `cli::cli_warn()`. We will also check if
the benchmark scales with the number of warnings in the function.

Based on this description, we will need the following packages: `cli` for its warning function,
[bench](https://cran.r-project.org/web/packages/bench/index.html) for measuring
the run times, `purrr`, `tidyr`, and `dplyr` for running the functions and
wrangling the results, and and `ggplot2` for visualizing the results.

```{r}
library(bench)
library(cli)
library(ggplot2)
library(purrr)
library(dplyr)
library(tidyr)
```

Let's define a function that takes an argument `n` for the number of warnings
to throw and `pkg` for the type of package to use.

```{r}
warn_Ntimes <- function(n, pkg) {
  warning_msg <- "x must be an integer"
  switch(pkg,
    base = for (i in 1:n) {
      warning(warning_msg)
    },
    cli = for (i in 1:n) {
      cli_warn(warning_msg)
    }
  )

> }

Author: Always good to catch out these patterns. Thanks, Change made here 468f20252d24d68fcdce8c2adc3c83bbd6ede74a.

```

Let's test our function to see if it works as expected.

```{r}
warn_Ntimes(3, "base")
warn_Ntimes(3, "cli")
```

Now, we'll consider scenarios where a function throws 1, 5, 10, 15, 20, and
100 warnings using base R and `cli`.

```{r warning=FALSE}
# Number of warnings to throw
warnings <- c(1, 5, 10, 15, 20, 100)
```

Let's run benchmarks over the different scenarios and store results
in a data.frame.

```{r warning=FALSE,message=FALSE}
bnmark_res <- bench::press(
  warnings = c(1, 5, 10, 15, 20, 100),
  bench::mark(
    cli = warn_Ntimes(warnings, "cli"),
    base = warn_Ntimes(warnings, "base")
    )
  )
```

## Results

It's time to explore the results of the data generated. We'll start by making
a boxplot of the median run times for the different scenarios.

Drum roll please...

```{r}

> autoplot(

Author: The default is 10k times. I thought to mention it but also wasn't sure if it was relevant. I was surprised at the dispersion in the base R results as well. Not sure what would explain that.

  bnmark_res,
  type = "boxplot"
  ) +
  labs(
    x = "Package",
    caption = "Functions throwing various numbers of warnings as indicated in the facet label"
  )
```

Finally, let's investigate the relative speeds of the two packages with
respect to the number of warnings. We will do this by dividing the median
benchmark of `cli::cli_warn()` by that of `base::warning()`.

```{r}
bnmark_relative_speeds <- bnmark_res |>
  select(expression, median, warnings) |>
  pivot_wider(
    names_from = expression,
    values_from = c(median)
    ) |> 
  mutate(
    relative_speed = cli/base
    )
bnmark_relative_speeds
```

As we can see in the new column `relative_speed`, `cli` is consistently
slower than base R and this seems not to

> scale linearly with the number of warnings thrown. Other benchmarking

Reviewer 3: The intention behind this statement wasn't clear to me - my expectation was that both {cli} and {base} times would scale linearly with $N$ warnings, which is met. If both scale roughly linearly, the relative difference should be roughly constant too as it is here. Is there a reason to expect otherwise?

It would be an issue if the relative difference _did_ scale, making {cli} unsuitable for high-performance applications with high-frequency output (e.g. perhaps some text-based interface for live monitoring?).

I would rephrase this to make it clear that there's an overhead to using {cli}, but that it is predictable, so there are unlikely to be nasty surprises (might be worth running for 10K warnings to be really sure).

Author: I've removed that section as it wasn't entirely useful. See 88c704d61df1a6d0fba81ab26c0aed018418795b.

experiments
with `cli` have revealed linear relationships between its speed and the number
of times it invokes other functions [^3]. The developers of `cli` have also

> conducted benchmarks of the `ansi_` family of functions in `cli` in comparison

Reviewer 2: Could add an asterisk (wildcard character) to `ansi_` (`ansi_*`) to show that this is a prefix for a family of functions.

Author: Thanks. I've made that change.

to base R and the `fansi` package. They find that `cli` is consistently slower
than base R, which corroborates the results of our experiment here.
Their benchmarks are available in the cli documentation [^4]. So, should we
be worried about the speed of `cli`? Well, it depends on the context. The
"R Packages" book by Hadley Wickham and Jenny Bryan has a section on
this where they suggest approaching such a decision from a holistic, balanced,

> and quantitative approach[^5]. We'll leave the reader to make their own decision

Reviewer 2: It might be worth fully citing the R packages book here. I don't know if we have a policy/preference on citations vs footnotes. I think citing would be a more clear sign of attribution for a published piece of work.

@Bisaloo any preferences?

Author: That's true.

Author: I've switched to using citations for books.

experiments. Here, we have demonstrated how benchmarking is one way to achieve
this. We show how a simple decision to use `cli::cli_warn()` to handle
warnings could come at the cost of a tiny loss in speed, which is worth
considering in its context.

The demonstration here can be extended to other design decisions such as
input checking, loops, object indexing, and so forth. We recommend benchmarking
as a way to help developers make design decisions. However, we also recommend
that developers consider the context of the optimisation in interpreting the results.

## Other R packages for benchmarking

* [microbenchmark](https://github.com/joshuaulrich/microbenchmark): an R
  package for comparing the execution time of R expressions.
* [rbenchmark](https://code.google.com/archive/p/rbenchmark/): an R package
  for benchmarking R code.
* [tictok](https://github.com/collectivemedia/tictoc): an R package to time
R functions
* [touchstone](https://github.com/lorenzwalthert/touchstone/tree/main): an R
  package for
  benchmarking of pull requests with statistical confidence.

[^1]: Benchmark (Wikipedia): <https://en.wikipedia.org/wiki/Benchmark_(computing)>

> [^2]:  Hadley Wickham's Advanced R chapter on Exceptions and Debugging: <http://adv-r.had.co.nz/Exceptions-Debugging.html>

Reviewer 2: I think this links to the old version of Advanced R. I'm not sure how much the content differs between versions, but it might be worth linking to the newer version of book: https://adv-r.hadley.nz/conditions.html. (Please check the new version contain the relevant info).

---
title: "Using benchmarking to guide the adoption of dependencies in R packages"
author:
  - name: "James Mba Azam"
    orcid: "https://orcid.org/0000-0001-5782-7330"

> date: last-modified

Reviewer: The `date` field will need to be updated to a static date before publishing to avoid re-publishing this post if we ever fix a typo or an URL in it.

Author: I've fixed that here acf2ede453b5ea607d98db1feaeb21863530e6ce.

> categories: [benchmarking, R, cli, R packages, software architecture]

Reviewer: ```suggestion
categories: [benchmarking, R, cli, R package, software architecture]
```

effort. For example, the `ggplot2` package is often adopted, instead of R's
inbuilt plotting functions, to make visualizations using its declarative
system and implementation of the Grammar of Graphics, and slick automation
across the board.

Sometimes, introducing a dependency might slow down existing processes. Hence,
it is important to consider the speed of a dependency's functions before
adopting it. Code optimisation is often reserved for heavy processes like data
processing steps. However, here, we will focus on lesser optimised processes,
and in particular, printing of texts in R.

For this demonstration, we will look at a situation where a package developer
is considering options for condition signalling beyond the functionalities
provided in base R. Condition signalling refers to the process of communicating
to a user that an issue has arisen during code execution [@wickham2019advanced].

Base R ships with functionalities for condition signalling. There is
`base::stop()` for signalling errors.  Errors refer to cases where something fatal
happens during code execution so that it halts, for example, when a function is
run without specifying the required arguments.

> ```{r}
> try(rnorm())
> ```

Reviewer: ````suggestion
```{r, error = TRUE}
rnorm()
```
````
